src/editor/Core/Filesystem.re

```diff
@@ -0,0 +1,301 @@
+type t('a) =
+  | Ok('a)
+  | Error(string);
+
+/** [on_success] executes [f] unless we already hit an error. In
+  that case the error is passed on. */
+let onSuccess = (t: t('a), f: 'a => t('b)) =>
+  switch (t) {
+  | Ok(x) => f(x)
+  | Error(str) => Error(str)
+  };
+
+/** [on_error] ignores the current error and executes [f]. If
+  there is no error, [f] is not executed and the result is
+  passed on. */
+let onError = (t, f) =>
+  switch (t) {
+  | Error(str) => f(str)
+  | Ok(x) => Ok(x)
+  };
+
+/**
+   Helper functions ============================================
+ */
+let _always = (_t, f) => f();
+
+/* This informs of an error and passes the error string wrapped in an Error to the next function*/
+let error = fmt => Printf.ksprintf(msg => Error(msg), fmt);
+
+let _inform = fmt => Printf.ksprintf(msg => Ok(msg), fmt);
+
+let return = x => Ok(x);
+
+let _fail = msg => Error(msg);
+
+/**
+   Infix Operators: These are essentially aliases to the
+   onError and onSuccess functions meaning that both of these can
+   this syntax allows us to use the monadic FS operators like
+   mkdir in a chainable fashion
+
+   mkdir
+    //= error("something went wrong")
+    >>= (_ => return())
+
+   in this example mkdir returns either an Ok(value) or an Error(value)
```

Comment:
Nice work thinking through all this, @Akin909 !

One thing we could consider is bringing in `ppx_let` which adds a helper for this monadic pattern: https://github.com/janestreet/ppx_let

Reply:
Cool @bryphe, I'm very curious about how these patterns can lead to dependable side-effect code. Although at first glance I have no idea what I'm looking at. I'll read it over and see if I can incorporate it here in a another PR.